### PR TITLE
CRI: Add Status into CRI.

### PR DESCRIPTION
**Pull request by Random-Liu**
For https://github.com/kubernetes/kubernetes/issues/35701.
Fixes https://github.com/kubernetes/kubernetes/issues/35701.

This PR added a `Status` call in CRI, and the `RuntimeStatus` is defined as following:

``` protobuf
message RuntimeCondition {
    // Type of runtime condition.
    optional string type = 1;
    // Status of the condition, one of true/false.
    optional bool status = 2;
    // Brief reason for the condition's last transition.
    optional string reason = 3;
    // Human readable message indicating details about last transition.
    optional string message = 4;
}

message RuntimeStatus {
    // Conditions is an array of current observed runtime conditions.
    repeated RuntimeCondition conditions = 1;
}
```

Currently, only `conditions` is included in `RuntimeStatus`, and the definition is almost the same with `NodeCondition` and `PodCondition` in K8s api.

@yujuhong @feiskyer @bprashanth If this makes sense, I'll send a follow up PR to let dockershim return `RuntimeStatus` and let kubelet make use of it.
@yifan-gu @euank Does this make sense to rkt?
/cc @kubernetes/sig-node

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35839)

### CODE REVIEW

**Comment by euank** [comment]
I don't think there's an issue with this from the rkt perspective. The same hostNetwork thing shhould behave pretty much identically here.

I do question the original choice of hostNetworkOnly being a condition though... thinking about it more, being able to reject a pod with a status like "networkNotReady" or "runtimeNotReady" as a reason would allow the kubelet to know to retry things in those cases and might be a more reliable way to handle these things.

**Comment by bprashanth** [comment]
rkt can just piggyback on kubenet's status(), which is a synthetic api to begin with (meaning it's not actually a part of CNI). 

Status does 2 things:     

1. checks if the cni binary exists at pluginDir
2. checks if the cni conf has been initialized with podCIDR

If either of those is false, the node remains notready. 
If --runtime=rkt --plugin=!kubenet, the rkt cri will have to implement a status() that effectively does the above.

**Comment by bprashanth** [comment]
@kubernetes/sig-network

File: pkg/kubelet/dockershim/remote/docker_service.go [hunk 1]
```diff
@@ -49,6 +49,10 @@ func (d *dockerService) Version(ctx context.Context, r *runtimeApi.VersionReques
 	return d.runtimeService.Version(r.GetVersion())
 }
 
+func (d *dockerService) Status(ctx context.Context, r *runtimeApi.StatusRequest) (*runtimeApi.StatusResponse, error) {
+	return nil, fmt.Errorf("not implemented")
```
Comment (bprashanth) [inline pkg/kubelet/dockershim/remote/docker_service.go]: Please add godoc. how are we planning to integrate this into the `defaultNodeStatusFuncs`? I'm guessing it needs to make its way into `runtimeState.networkErrors`?

File: pkg/kubelet/api/v1alpha1/runtime/api.pb.go [hunk 2]
```diff
@@ -2603,6 +2607,109 @@ func (m *UpdateRuntimeConfigResponse) String() string            { return proto.
 func (*UpdateRuntimeConfigResponse) ProtoMessage()               {}
 func (*UpdateRuntimeConfigResponse) Descriptor() ([]byte, []int) { return fileDescriptorApi, []int{71} }
 
+// RuntimeCondition contains condition information for the runtime.
+// There are 2 kinds of runtime conditions:
+// 1. Required condtitions: Conditions are required for kubelet to work
+// properly, including:
+//   * RuntimeReady: RuntimeReady means the runtime is up and ready to accept
+//   basic containers.
+//   * NetworkReady: NetworkReady means the runtime network is up and ready to
```
Comment (bprashanth) [inline pkg/kubelet/api/v1alpha1/runtime/api.pb.go]: so if any required condition is unmet the node remains notready (please elaborate comment)? 
we will probably have to conditionally start hostnetwork pods if NetworkReady == false but everything else checks out.

File: pkg/kubelet/dockershim/remote/docker_service.go (hunk 1 above)
Comment (yujuhong) [inline pkg/kubelet/dockershim/remote/docker_service.go]: Yes, we probably need to change [`syncNetworkStatus`](https://github.com/kubernetes/kubernetes/blob/2941aea88e80e94f2698241a4ac9925145664470/pkg/kubelet/kubelet_network.go#L193) to pipe this information.

**Review by yujuhong — CHANGES_REQUESTED** [review]
Need to update the internal interface in `pkg/kubelet/api/services.go` and add an implementation stub in `dockershim/docker_service.go`

**Comment by Random-Liu** [comment]
Will finish the following part today.

File: pkg/kubelet/api/v1alpha1/runtime/api.pb.go (hunk 2 above)
Comment (Random-Liu) [inline pkg/kubelet/api/v1alpha1/runtime/api.pb.go]: ACK

**Comment by feiskyer** [comment]
> Need to update the internal interface in pkg/kubelet/api/services.go and add an implementation stub in dockershim/docker_service.go

Besides, also add `Status()` calling at `pkg/kubelet/remote` and `pkg/kubelet/api/testing`.

File: pkg/kubelet/dockershim/remote/docker_service.go (hunk 1 above)
Comment (bprashanth) [inline pkg/kubelet/dockershim/remote/docker_service.go]: Shouldn't the kubelet parses out network/runtime status directly from the response of this rpc in syncRuntimeStatus?

**Comment by Random-Liu** [comment]
Addressed comments and implement the initial version of `Status`.

I'll send another PR to expose RuntimeReady and NetworkReady from runtime interface https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/container/runtime.go#L73.

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

File: pkg/kubelet/api/testing/fake_runtime_service.go [hunk 3]
```diff
@@ -50,10 +50,21 @@ type FakeRuntimeService struct {
...
 
+	Errors     map[string]error
+	FakeStatus *runtimeApi.RuntimeStatus
 	Containers map[string]*FakeContainer
 	Sandboxes  map[string]*FakePodSandbox
 }
 
+// SetFakeRuntime is a simple function to set arbitrary field in the fake runtime
+// with lock.
+func (r *FakeRuntimeService) SetFakeRuntime(set func(*FakeRuntimeService)) {
```
Comment (Random-Liu) [inline pkg/kubelet/api/testing/fake_runtime_service.go]: Notice that this is not used, but will be used in the future unit test.

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

File: pkg/kubelet/api/testing/fake_runtime_service.go [hunk 4]
```diff
@@ -50,10 +50,21 @@ type FakeRuntimeService struct {
 
 	Called []string
 
+	Errors     map[string]error
```
Comment (feiskyer) [inline pkg/kubelet/api/testing/fake_runtime_service.go]: what is this for?

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go [hunk 5]
```diff
@@ -261,13 +261,23 @@ func (m *kubeGenericRuntimeManager) APIVersion() (kubecontainer.Version, error)
...
+	networkReady := getRuntimeCondition(status, internalApi.NetworkReady)
+	if networkReady == nil || !networkReady.GetStatus() {
+		return fmt.Errorf("runtime network not ready: reason: %q, message: %q",
+			networkReady.GetReason(), networkReady.GetMessage())
+	}
+	runtimeReady := getRuntimeCondition(status, internalApi.RuntimeReady)
+	if runtimeReady == nil || !runtimeReady.GetStatus() {
+		return fmt.Errorf("runtime not ready: reason: %q, message: %q",
+			runtimeReady.GetReason(), runtimeReady.GetMessage())
 	}
```
Comment (feiskyer) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: nit: do we need to report node events if there are other not ready conditions?

File: pkg/kubelet/api/testing/fake_runtime_service.go (hunk 4 above)
Comment (Random-Liu) [inline pkg/kubelet/api/testing/fake_runtime_service.go]: This is for injecting errors.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 5 above)
Comment (Random-Liu) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: The code is temporary, I'll send another PR/commit soon to really use the conditions. I'll deal with then.

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

File: pkg/kubelet/kuberuntime/kuberuntime_manager.go (hunk 5 above)
Comment (feiskyer) [inline pkg/kubelet/kuberuntime/kuberuntime_manager.go]: LGTM

**Comment by Random-Liu** [comment]
PR is ready for review.

**Comment by Random-Liu** [comment]
@k8s-bot cri test this.

**Comment by feiskyer** [comment]
LGTM. needs @yujuhong take another look, and confirm review changes.

File: pkg/kubelet/kubelet.go [hunk 6]
```diff
@@ -1899,10 +1893,40 @@ func (kl *Kubelet) PLEGHealthCheck() (bool, error) {
...
 func (kl *Kubelet) updateRuntimeUp() {
-	if err := kl.containerRuntime.Status(); err != nil {
+	s, err := kl.containerRuntime.Status()
+	if err != nil {
 		glog.Errorf("Container runtime sanity check failed: %v", err)
 		return
 	}
+	// Only check specific conditions when runtime integration type is cri,
+	// because the old integration doesn't populate any runtime condition.
+	if kl.kubeletConfiguration.ExperimentalRuntimeIntegrationType == "cri" {
```
Comment (yujuhong) [inline pkg/kubelet/kubelet.go]: Never mind. I decided to send in #36201 to enforce that all integration using CRI should specify --experimental-runtime-integration-type=cri for now to simplify the code.

**Review by yujuhong — APPROVED** [review]
LGTM.

**Comment by Random-Liu** [comment]
@yujuhong Add the `networkPlugin=nil` back.

**Comment by Random-Liu** [comment]
Just rebase, apply LGTM again.

**Pull request merged**